### ISSUE TITLE
Simplify Dynamic DNS configuration

### ISSUE THREAD

**Issue by SunilMohanAdapa** [freedombox/Plinth#390]
Current there is a lot of code in the Dynamic DNS action module to handle the case of NAT detection and based on that, running the ez-ipupdate either from cron or as daemon.  Explore ways to simplify the logic.

**Comment by steglicd** [comment]
@SunilMohanAdapa I agree with you. The action module only was meant to be a wrapper script for ez-ipupdate. I would rewrite the whole thing, using python. Because the logic behind the script is a bit more than a simple wrapper script and therefore should not be written in bash.

Did you already started refactoring?

**Comment by SunilMohanAdapa** [comment]
On 02/01/2016 01:01 PM, Daniel Steglich wrote:
> @SunilMohanAdapa <https://github.com/SunilMohanAdapa> I agree with you.
> The action module only was meant to be a wrapper script for ez-ipupdate.
> I would rewrite the whole thing, using python. Because the logic behind
> the script is a bit more than a simple wrapper script and therefore
> should not be written in bash.

Rewrite in Python would be awesome.

The WAN IP address that we detect in case we are behind NAT does not
seem to go as input the ez-ipupdate anywhere.  Does this mean
ez-ipupdate can actually work for situations behind NAT but does not
give us any indication of the NAT status.

If there is a way to avoid running cron job as wrapper, that would
reduce the amount of work we have do making things very simple.  If the
daemon is exiting frequently we could write a systemd service to auto
restart it every few minutes.  Do you have any ideas to simplify?

> 
> Did you already started refactoring?

I did not.  I am still occupied with the review tasks.  Would be nice if
you are able make time for it.

-- 
Sunil

**Comment by steglicd** [comment]
> The WAN IP address that we detect in case we are behind NAT does not seem to go as input the ez-ipupdate anywhere. Does this mean ez-ipupdate can actually work for situations behind NAT but does not give us any indication of the NAT status.
The ez-ipupdate tool supports two modes:
1. daemon mode -> will monitor IP address changes of the given interface, will only work without NAT (direct connection to the internet)
2. single update call -> Will do a single update, will not pass a IP address to server, the server will use the source IP of the request

For the second method, we need to check for WAN IP changes and/or call ez-ipupdate tool periodically.

> If there is a way to avoid running cron job as wrapper, that would reduce the amount of work we have do making things very simple. If the daemon is exiting frequently we could write a systemd service to auto restart it every few minutes. Do you have any ideas to simplify?
The Daemon will only exit in case 2, but a systemd service which will start the daemon every few minutes will not replace all the script magic (check if IP has changed, update the ez-ipupdate config file, call the daemon).

Also, if a update URL (so no GnuDIP protocol) is used, the daemon will not  be started in any case. We need to do a HTTP GET, which is not done by ez-ipupdate.

I would like to work on this issue. I think I will draw a chart first, so we can see all the use cases and discuss them. OK @SunilMohanAdapa  ?

**Comment by steglicd** [comment]
![dyndns](https://cloud.githubusercontent.com/assets/7570118/12733263/27ea23c6-c93a-11e5-99cf-8cc7ff72d95b.png)


I made a drawing of what the dynamic DNS (action) module is currently doing.
As @SunilMohanAdapa stated, this is very complex and it would be better to simplify.

The main problem is, that this module is a bit different to other FB modules. Other modules are mainly 3rd party software components which only needs to be launched by plinth. But ez-ipupdate is missing a lot of features (like generic update URL), so the action script was extended to offer this functionality. In the drawing you can see "ez-ipudate package magic (debian)", "plinth magic" and "dynamic DNS magic". Where other modules will only have the plinth- and the package magic.

Any comments / suggestions on how to simplify (beside the fact, that it should be written in python)?

**Comment by SunilMohanAdapa** [comment]
@steglicd, Thank you for taking the time to explain and create the flow diagram.  I understand now why we are doing what we are doing.

One potential simplification I can think of is to do the 'no NAT' case same as the 'NAT' case.  That is use the cron job with 'update' command and nothing else.
- Install a cronjob permanently
- Never run ez-ipupdate in daemon mode
- Handle no NAT case in update command (if current DNS entry is one of our ips, skip. Otherwise run NAT case or direct update)

With this, there won't be need to install/uninstall cron job, enable/disable service, move configuration files look for running processes etc.

Another possibility is to look for more mature client that won't require us to implement a lot of functionality.  Even if it is not in Debian, this time, we can work on packaging it. Perhaps ddns-scripts from OpenWrt?  ddclient? I suppose you have done more research in the area.

**Comment by SunilMohanAdapa** [comment]
If exploring other Dynamic DNS clients seems excessive, we can at least convert the script to Python while applying the simplification above.

**Comment by steglicd** [comment]
Sorry, I was a bit lazy in documenting progress.
So actually I did not find other ddns clients which offers the same features.

I.e. as far as I can see from ddclient documentation it does not support GnuDIP. One may say, so ignore GnuDIP. But I think the protocol is much safer than HTTP GET.

Also ddclient does not  support generic update URL. The have multiple update protocols like "dyndns2", "dyndns1", "easydns", "dslreports1", "hammernode1" ... 

I think there is a specifc URI behind each of them (so: like update URL) but who knows which URI is associated to which protocol? And what if I want to use a different one?

Same on openwrt. And openwrt is not available as debian package.

OK, what now? @SunilMohanAdapa the complexity of the dyndns update code is huge. Do we want to maintain this inside of FB? I could convert the script to python. Do we want to have this inside FB? I also could create a new project on github and we include this from FB?
